### DISCUSSION TITLE
GSoC 2022: On-disk encryption

### DISCUSSION THREAD

**Discussion by solokyo** [pgmoneta/pgmoneta#56] [Ideas]
Hi,
This is Jichen Xu, glad to be selected as a GSoC contributor this year.
Here is my proposal for GSoC 2022:

## Project Description
This project aims to add support for on-disk encryption of backups.
These are some tasks that need to be completed:
* Create an internal API for encryption / decryption
* Create an implementation using AES-256

## Approach
Investigate different variants of the AES algorithm, get a deeper understanding in the advantages and disadvantages of different key length(128, 192, 256 bits) and different modes(CBC ECB CTR OCB CFB).

After investigation, determine which key length and mode(s) will be included, then implement a public interface (aes.h) along with an implementation(aes.c). Alter the configuration, add options for switching between plaintext/cipher mode, choose the key length and AES encryption mode.

Some refactoring may be needed to make sure newly added functionality is compatible with the current codebase. Update the documentation and/or tutorial about the configuration of encryption after coding and testing.

## Deliverables
* Add new options to configuration for encryption-related setting
* Implement encryption and decryption 
* Make cipher mode compatible with plaintext mode
* Update the documentation and/or tutorial about configuration (if needed)

For more details, check my proposal on [Google Docs](https://docs.google.com/document/d/1vdgPY5wvhjhrX9aSUw5mTDOnXwEQNcr4cxfzuSHMWlg/).

:smiley: Any suggestions and discussions are welcomed.

**Comment by jesperpedersen** [comment]
Welcome !

I think the important part of this project is to get a deeper understanding of the encryption algorithms that we can use - and will support - but more importantly which algorithms we won't support and why. I expect that we will stick with AES as a baseline for this work, but what else is out there ? We can document this in the forum.

There is already AES-256 functionality inside pgmoneta, but it needs to refactored out as you say such that it can be used generally across the code base, so `aes.h` and `aes.c` is a good idea.

As encryption is optional and will add an overhead to the backup I think an idea would be to create a table / graph that shows the time needed for different data sizes per algorithm. F.ex. pgbench could be used to generate these databases (1, 5, 10, ... Gb). This information should likely go into a separate document (`ENCRYPTION.md`) that explain the supported algorithms and their properties. Maybe export restrictions plays a role in our selection ? Can we use SIMD, or it is transparent, or N/A for this ? 

I think for `CONFIGURATION.md` is likely enough to do document the basic setting, like

```
encryption = none (default) | aes-128 | aes-192 | aes-256
```

but we can look at that later.

An idea could be to start to investigate the different algorithms and post a pros/cons list here in the forum.

**Comment by solokyo** [comment]
First, my apologies for the late reply. My midterm exam is really exhausted.

</br>
Cryptographic algorithms can be roughly categorized into two classes: symmetric and asymmetric.

Asymmetric algorithms are slower than symmetric algorithms because it requires more mathematics. How slow exactly? [Here](http://bench.cr.yp.to/results-encrypt.html) is the benchmark. In practice, we're using Hybrid Cryptosystem.

For symmetric algorithms, they can also be separated into two categories: block ciphers and stream ciphers. [Here](https://security.stackexchange.com/questions/334/advantages-and-disadvantages-of-stream-versus-block-ciphers) is a good post to tell the difference. Both categories have a lot of implementations out there. For stream ciphers, [Here](http://bench.cr.yp.to/primitives-stream.html) is a list with a benchmark. While I am reading [the paper](https://eprint.iacr.org/2013/295.pdf) trying to get understand each different block cipher and it's benchmarks, I realized that most modern processors support [AES-NI](https://en.wikipedia.org/wiki/AES_instruction_set) (You can verify your pc by ```grep aes /proc/cpuinfo```). Hardware acceleration made AES outperform other algorithms. In my opinion, we don't need to look at other algorithms. So we focus on AES.
</br>
Now the problem is decide which AES mode(s) will be included in our code. [This post](https://stackoverflow.com/questions/1220751/how-to-choose-an-aes-encryption-mode-cbc-ecb-ctr-ocb-cfb) gives a good explanation of each mode. I would suggest including at least two modes:
* CBC: It is the most commonly used mode and considered safe, actually it is already in our code base.
* CTR: The main drawback of CBC mode is that encryption is sequential (Decryption can be parallelized). If our user has a strict performance requirement, CTR is an option for them because CTR can do both encryption and decryption in parallel.

If the authenticity is our or our user's concern, we may include Authenticated Encryption (AES-GCM) as well. [See this](https://crypto.stackexchange.com/questions/12178/why-should-i-use-authenticated-encryption-instead-of-just-encryption) for a detailed explanation about When and why to use Authenticated Encryption.
</br>
For key length, it is easy to integrate all three types of key length (128, 192, 256) in our code, we can leave our user to decide. However, I would suggest using 128 as the default key length for the following two reasons:

1. 128 doing 10 rounds, 192 doing 12 rounds, 256 doing 14 rounds of XOR operation. That is 20% and 40% extra workload for extra safety. 
2. AES-128 is far from broken. [See this](https://crypto.stackexchange.com/questions/77000/is-128-bit-security-still-considered-strong-in-2020-within-the-context-of-both).

</br>

>  I think an idea would be to create a table / graph that shows the time needed for different data sizes per algorithm. F.ex. pgbench could be used to generate these databases (1, 5, 10, ... Gb). 

I don't know if the type of data (plaintext) would affect encryption and decryption performance. If not, we may use any data to run a benchmark. The performance of encryption and decryption is more related to hardware specifications. Since our user's hardware can be very different, I suggest running [openssl speed](https://www.openssl.org/docs/man1.1.1/man1/openssl-speed.html) on different machines to get our own benchmark, or we can just refer to other people's benchmark.
[Here](https://openwrt.org/docs/guide-user/perf_and_log/benchmark.openssl#openssl_benchmarks) is a CBC benchmark on different machines, and [here](https://www.highgo.ca/2019/08/22/the-performance-test-on-the-aes-modes/) is a benchmark of each AES modes on the same machine.

</br>

> Maybe export restrictions plays a role in our selection?

I am not sure what is your concern. I know some of the algorithms such as AES-OCB is patented, I will make sure that any algorithm we included won't cause license or patent problem.

</br>

>  Can we use SIMD, or it is transparent, or N/A for this ?

This part is also a little bit ambiguous to me. I am thinking something like GPU acceleration, am I on the right track? If so I will continue to investigate on this.

**Reply by jesperpedersen** [reply]
Thanks for the detailed description !

I agree that we should focus on AES w/ 128/192/256 + CBC|CTR. We could make an alias of `aes-256` to be the same as `aes-256-cbc`. I think it is ok to have `aes-256` being the default - if people really understand the AES algorithms based on the documentation they can change it to `aes-128` if they want the speed.

As for the benchmark we just list the basic machine information - it is just to show the graphs. If there is an easy way to include a benchmark script we can do so in `/contrib`.

For SIMD I was mostly thinking about if there is a special flag for OpenSSL for it - otherwise we can circle back to it once the functionality has been implemented.

Great work !
